apps/aevm/src/aevm_ae_primops.erl

```diff
@@ -176,109 +176,123 @@ oracle_ttl_t() ->
+                fun(ChainAPI, ChainState) ->
+                    case ChainAPI:oracle_register(Tx, to_sign(Sign0), ChainState) of
                         {ok, <<OKey:256>>, ChainState1} -> {ok, OKey, ChainState1};
                         {error, _} = Err                -> Err
                     end
                 end,
-            DynCost = state_gas_cost(oracle_registration, DeltaTTL),
-            ?TEST_LOG("~s computed gas cost ~p from relative TTL ~p", [?FUNCTION_NAME, DynCost, DeltaTTL]),
-            {ok, DynCost, fun() -> call_chain(Callback, State) end}
+            SizeGas = size_gas(Tx),
+            StateGas = state_gas(oracle_registration, DeltaTTL),
+            DynGas = SizeGas + StateGas,
+            ?TEST_LOG("~s computed gas ~p - size gas: ~p, state gas: ~p (relative TTL: ~p)",
+                        [?FUNCTION_NAME, DynGas, SizeGas, StateGas, DeltaTTL]),
+            {ok, DynGas, fun() -> call_chain(Callback, Chain) end}
     end.
 
-oracle_call_query(Value, Data, State) ->
+oracle_call_query(Value, Data, #chain{api = API, state = State} = Chain) ->
     [Oracle]  = get_args([word], Data),  %% We need the oracle address before we can decode the query
-    OracleKey = <<Oracle:256>>,
-    case call_chain1(fun(API, ChainState) -> API:oracle_query_format(OracleKey, ChainState) end, State) of
-        {ok, QueryType} ->
-            ArgumentTypes = [word, QueryType, oracle_ttl_t(), oracle_ttl_t()],
-            [_Oracle, Q, QTTL, RTTL] = get_args(ArgumentTypes, Data),
-            case chain_ttl_delta(QTTL, State) of
+    case API:oracle_query_format(<<Oracle:256>>, State) of
+        {ok, QFormat} ->
+            ArgumentTypes = [word, QFormat, oracle_ttl_t(), oracle_ttl_t()],
+            [Oracle, Q, QTTL, RTTL] = get_args(ArgumentTypes, Data),
+            case chain_ttl_delta(QTTL, Chain) of
                 {error, _} = Err -> Err;
                 {ok, DeltaQTTL = {delta, _}} ->
+                    {ok, Tx} = API:oracle_query_tx(<<Oracle:256>>, Q, _QFee=Value, QTTL, RTTL, State),
                     Callback =
-                        fun(API, ChainState) ->
-                            case API:oracle_query(OracleKey, Q, _QFee=Value, QTTL, RTTL, ChainState) of
+                        fun(ChainAPI, ChainState) ->
+                            case ChainAPI:oracle_query(Tx, ChainState) of
                                 {ok, <<QKey:256>>, ChainState1} -> {ok, QKey, ChainState1};
                                 {error, _} = Err                -> Err
                             end
                         end,
-                    DynCost = state_gas_cost(oracle_query, DeltaQTTL),
-                    ?TEST_LOG("~s computed gas cost ~p from relative TTL ~p", [?FUNCTION_NAME, DynCost, DeltaQTTL]),
-                    {ok, DynCost, fun() -> call_chain(Callback, State) end}
+                    SizeGas = size_gas(Tx),
```

Comment:
Again the size of what we put in the state should matter, the size of the Tx is already charged for in #1674